doc/src/manual/arrays.md

```diff
@@ -358,6 +336,141 @@ julia> x
+    * Empty arrays like `[]`, which select no elements
+    * `Range`s of the form `a:c` or `a:b:c`, which select contiguous or strided subsections from `a` to `c` (inclusive)
+    * Any custom array of scalar indices that is a subtype of `AbstractArray`
+    * Arrays of `CartesianIndex{N}` (see below for more details)
+3. An object that represents an array of scalar indices and can be converted to such by [`to_indices`](@ref). By default this includes:
+    * [`Colon()`](@ref) (`:`), which represents all indices within an entire dimension or across the entire array
+    * Arrays of booleans, which select elements at their `true` indices (see below for more details)
+
+#### Cartesian indices
+
+The special `CartesianIndex{N}` object represents a scalar index that behaves
+like an `N`-tuple of integers spanning multiple dimensions.  For example:
+
+```jldoctest cartesianindex
+julia> A = reshape(1:32, 4, 4, 2);
+
+julia> A[3, 2, 1]
+7
+
+julia> A[CartesianIndex(3, 2, 1)] == A[3, 2, 1] == 7
+true
+```
+
+Considered alone, this may seem relatively trivial; `CartesianIndex` simply
+gathers multiple integers together into one object that represents a single
+multidimensional index. When combined with other indexing forms and iterators
+that yield `CartesianIndex`es, however, this can lead directly to very elegant
+and efficient code. See [Iteration](@ref) below, and for some more advanced
+examples, see [this blog post on multidimensional algorithms and
+iteration](http://julialang.org/blog/2016/02/iteration).
+
+Arrays of `CartesianIndex{N}` are also supported. They represent a collection
+of scalar indices that each span `N` dimensions, enabling a form of indexing
+that is sometimes referred to as pointwise indexing. For example, it enables
+accessing the diagonal elements from the first "page" of `A` from above:
+
+```jldoctest cartesianindex
+julia> page = A[:,:,1]
+4×4 Array{Int64,2}:
+ 1  5   9  13
+ 2  6  10  14
+ 3  7  11  15
+ 4  8  12  16
+
+julia> page[[CartesianIndex(1,1),
+             CartesianIndex(2,2),
+             CartesianIndex(3,3),
+             CartesianIndex(4,4)]]
+4-element Array{Int64,1}:
+  1
+  6
+ 11
+ 16
+```
+
+This can be expressed much more simply with [dot broadcasting](@ref man-vectorized)
+and by combining it with a normal integer index (instead of extracting the
+first `page` from `A` as a separate step). It can even be combined with a `:`
+to extract both diagonals from the two pages at the same time:
+
+```jldoctest cartesianindex
+julia> A[CartesianIndex.(indices(A, 1), indices(A, 2)), 1]
+4-element Array{Int64,1}:
+  1
+  6
+ 11
+ 16
+
+julia> A[CartesianIndex.(indices(A, 1), indices(A, 2)), :]
+4×2 Array{Int64,2}:
+  1  17
+  6  22
+ 11  27
+ 16  32
+```
+
+!!! warning
+
+    `CartesianIndex` and arrays of `CartesianIndex` are not compatible with the
+    `end` keyword to represent the last index of a dimension. Do not use `end`
+    in indexing expressions that may contain either.
```

Comment:
That would be an improvement, and make the subject more obvious. Would be most explicit but a bit repetitive to just say what's being referred to.